### PR TITLE
8332124: Jcmd should recognise options that look like requests for help

### PR DESCRIPTION
**Pull request by SoniaZaldana**
Hi all, 

This PR addresses [8332124](https://bugs.openjdk.org/browse/JDK-8332124) enabling jcmd to recognize options that look like help. 

Testing: 
- [x] Added test case passes. 

Thanks, 
Sonia

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8332124](https://bugs.openjdk.org/browse/JDK-8332124): Jcmd should recognise options that look like requests for help (**Enhancement** - P4)

### Reviewers
 * [Kevin Walls](https://openjdk.org/census#kevinw) (@kevinjwalls - **Reviewer**)
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**) 🔄 Re-review required (review applies to [56203e57](https://git.openjdk.org/jdk/pull/19776/files/56203e57372d940875f8f620fd35e842e4948391))

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19776/head:pull/19776` \
`$ git checkout pull/19776`

Update a local copy of the PR: \
`$ git checkout pull/19776` \
`$ git pull https://git.openjdk.org/jdk.git pull/19776/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19776`

View PR using the GUI difftool: \
`$ git pr show -t 19776`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19776.diff">https://git.openjdk.org/jdk/pull/19776.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19776#issuecomment-2187326751)

### CODE REVIEW

**Review by kevinjwalls — COMMENTED** [review]
I could benefit from this, I think it's natural to sometimes want to check the help after you've started typing, so I like the idea.  But can it actually clash with any other usage?

e.g. "help" can be a filename.  We are making the filename "help" unusable.  Maybe not a big deal to most people, but we do need to make this clear.


jcmd PID Thread.dump_to_file help

...used to create a file "help", and now does not?

Compiler.directives_add takes a filename and would accept "help".
GC.heap_dump used to create a dump file called "help".

jcmd 7537 VM.set_flag help
..could have been a clash if we had such a VM flag (not likely...).


I don't mean to labour artificial or uncommon cases, but we do want this to be clearly specified.  Maybe this is still OK with the right hints (somewhere?) about the filenames that are now not possible to read or create.


Any thoughts on if
jcmd PID command -help
or 
jcmd PID command --help

..should work, or could work?  Additionally, or instead?  Maybe they are better, as users relying on files called "-help" would be really uncommon (and would be inviting problems!).


Good to ignore the trailing whitespace.
Not obvious whether trailing extra text should be ignored or cause an error.
e.g.
jcmd 7537 VM.stringtable help nonsense
7537:
java.lang.IllegalArgumentException: Unknown argument 'help' in diagnostic command.

..where I guess it's not the "help" which is unknown.  Is the user better served by just giving the help and ignoring the additional "nonsense"?

**Comment by tstuefe** [comment]
@SoniaZaldana, thank you for addressing this, this has been bugging me for a long time (hence the issue).

@kevinjwalls

> I could benefit from this, I think it's natural to sometimes want to check the help after you've started typing, so I like the idea. But can it actually clash with any other usage?

good point

> Any thoughts on if jcmd PID command -help or jcmd PID command --help

Yes, something like that seems the only obvious solution. 

Regrettably, we follow no standard at all for jcmd argument formats. I have been wishing for years that we would follow something consistent, e.g. the [Posix utility argument syntax](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap12.html#tag_12_01). Or however the Gnu variant is called with their double-dash arguments.

Something like "--help" would probably not clash with anything, e.g. also not with a future introduction of Posix utility argument syntax (dash-multiple-flags). Or, something like "-h". But we would have to examine all commands.

Sigh. Maybe adding "-h" to each command would have been the better way, if we need to examine every command anyway to check for compat problems. Or, to the command base classes. Lets see what other reviewers think.

> 
> ..where I guess it's not the "help" which is unknown. Is the user better served by just giving the help and ignoring the additional "nonsense"?

Yes. We should ignore trailing arguments after --help

**Comment by kevinjwalls** [comment]
Thanks Thomas -

Yes I could see all of "jcmd PID XX.command -help" and -h and --help being recongised, and giving help, and ignoring anything else on the line after that.

We dont have any "--" prefixes in use, but if we can recognise it and show the help that could still be good and I don't think it can clash.

Hopefully this kind of thing can still be done without changing every implementing class. 8-)

Yes, standards...  Agreed it's a concern!... 
FYI while looking into this I was looking through the commands, and have also started:
8335154: jcmd VM.classes -verbose=false does not set verbose to false
8335169: jcmd should standardize on "-verbose" not "verbose" (ClassLoaderHierarchyDCmd and JFR.check are different to the rest)

**Comment by SoniaZaldana** [comment]
Hi all, 

Thanks for the comments. Some thoughts: 

>  what about '*jcmd <pid> help <cmd>*'
> not a great deal of additional code for the 'help' jcmd to delegate to
> the target jcmd?

```jcmd <pid> help <cmd>``` is already supported. This patch aims to rearrange the arguments provided to leverage what you suggested  (i.e. ```jcmd <pid> <cmd> help``` -> ```jcmd <pid> help <cmd>```). Not sure if I misunderstood your message though :-) 

Regarding the current state of the patch,
> e.g. "help" can be a filename. We are making the filename "help" unusable. Maybe not a big deal to most people, but we do need to make this clear.

This is a really good point and in my opinion the strongest argument to use `--help` or `-help` instead. My only pushback would be whether it would be a bit inconsistent to make it so that the dashes need to be specified for help in this case ```jcmd <pid> <cmd> -help```, but not in this one ```jcmd <pid> help <cmd>```.

Unless we plan to rework “help” in general and make it so that you always need the dashes regardless of how it gets invoked. 

> 
> jcmd 7537 VM.stringtable help nonsense
> 7537:
> java.lang.IllegalArgumentException: Unknown argument 'help' in diagnostic command.
> ..where I guess it's not the "help" which is unknown. Is the user better served by just giving the help and ignoring the additional "nonsense"?

Sure, that makes sense to me. I can make an update to ignore additional nonsense once we agree on the other points. 

As a final question, if we make it so we need the dashes to enable help, would we need to document this anywhere or do we assume it's self-explanatory as it follows the Posix utility argument syntax that Thomas linked?

Thanks for all the feedback!

**Comment by kevinjwalls** [comment]
Hi Sonia --

I think this JBS entry exists to try and help out users who go fishing for help.  They are off the documented path, but we can try and do what they want.

I don't think we should document this attempt to help people out (partly in case it does cause any conflicts later).

We we can't help people who type "help" where there should be a filename, and we don't want to make anything incompatible or change the existing help command.

So if you type
 jcmd PID command -help 
or even 
jcmd PID command --help 

..then we could help you by showing the help text.  Agreed on ignoring anything additional after jcmd PID command -help (or --help).

I don't think there is a problematic inconsistency with the existing "help" command, and these new help options.  Commands don't have any - prefix, so the help command doesn't.  But arguments to a command can have - prefixes.

It is obviously inconsistent to recognise posix-style "--help", but it's clearly a user asking for help and I have no problem with it - although do wait and see if anybody else objects. 8-)


This change prohibits a DCmd from implementing a "-help" argument.  That is not a problem today, we don't have any that do that (and it seems too much right now to inspect the DCmd to see if it does have a "-help" argument).

BTW in your current src/hotspot/share/services/diagnosticFramework.cpp:405 it would be great to not be specific about exactly one space before "help/-help/--help".

**Comment by tstuefe** [comment]
I would not use help with one dash. Reason: if we ever wanted to switch to posix conform arguments, this would translate to four flags, h+e+l+p. 

I would either use -h or --help, preferably the former.

**Comment by tstuefe** [comment]
I looked through all our commands, I did not see a single one that conflicted with "-h". So that is possible.

**Comment by SoniaZaldana** [comment]
One question. With the current implementation, this is the check in place to offer help: 
```strncmp(args, " -help", 6) == 0 || strncmp(args, " --help", 7)```

I could change this to do `strncmp(args, " -h", 3)` instead but this is a bit problematic as it would block any future flags that start out with `-h` from working.

I've been giving this some thought on how to implement a restrictive way to check only for the arguments `-h` or `-help` while also allowing for other arguments to be passed afterwards that should be ignored. 

I thought about perhaps using some type of regex matching with ```std:regex``` but hotspot doesn't allow the use of global operators new and delete.  I don't want to overengineer this piece of logic so I was wondering if there was a regex matching utility that I could leverage in HotSpot? 

The alternative would be to make the check more restrictive and not allow for arguments after `-h` has been issued i.e. `strcmp(args, "-h") == 0`. 

Thanks for your help!

**Comment by tstuefe** [comment]
Hi @SoniaZaldana 

sorry for the delay, I'm snowed in atm.

> One question. With the current implementation, this is the check in place to offer help: `strncmp(args, " -help", 6) == 0 || strncmp(args, " --help", 7)`
> 
> I could change this to do `strncmp(args, " -h", 3)` instead but this is a bit problematic as it would block any future flags that start out with `-h` from working.

Well, we restrict future compatibility no matter what we do. 

> 
> I've been giving this some thought on how to implement a restrictive way to check only for the arguments `-h` or `-help` while also allowing for other arguments to be passed afterwards that should be ignored.
> 
> I thought about perhaps using some type of regex matching with `std:regex` but hotspot doesn't allow the use of global operators new and delete.

God no :)

> I don't want to overengineer this piece of logic so I was wondering if there was a regex matching utility that I could leverage in HotSpot?

No need, you can do this with normal C tools.

with p pointing to start of arguments:
- forward p and walk all spaces, break out for \0 
- strncmp with "-h", size 2
- if follow-up char is either blank or \0, its a help command. If we agree to just ignore follow up arguments, you can now just rebuild the command as "help commandname".
- bonus for writing a warning if there are arguments after help

> 
> The alternative would be to make the check more restrictive and not allow for arguments after `-h` has been issued i.e. `strcmp(args, "-h") == 0`.
> 
> Thanks for your help!

Cheers

**Comment by kevinjwalls** [comment]
Hi,

1.
The JBS issue and PR title need an update. Maybe something like:

  Jcmd should recognise options that look like requests for help


2.
I would encourage recognising "-help" in this change.  This is contrary to Thomas' suggestion, and I see the clash with posix style --options, but we already have many many single-minus-prefixed options.  This change as I see it is an undocumented attempt to help lost users.  If we start a migration to posix-style, much effort will be needed so this clash, which could be fixed then, will seem insignificant.


3.
I want to mention the alternative to this change: print the command help whenever showing an error.  The downside would be that it can be quite long, and maybe you don't want to see several screens of help when you mistype one option to jcmd JFR.start...
I think we should complete what we have here already, but we should note that there are alternatives.


4.
sun/tools/jcmd/TestJcmdSubcommandHelp.java
The summary at line 28 needs updating, whether you add "-help" or not.

Would be great to test "jcmd VM.command -hello" and check it does not show help.  I added one and tried it out, will paste it in here somehow.


Thanks!

**Comment by kevinjwalls** [comment]
Additional test possibility.

```
*** orig.txt    2024-07-03 13:08:46.460514793 +0100
--- test/jdk/sun/tools/jcmd/TestJcmdSubcommandHelp.java 2024-07-03 12:50:48.324022729 +0100
***************
*** 39,48 ****
--- 39,49 ----
  public class TestJcmdSubcommandHelp {

      private static final String HELP_ONE_DASH = "-h";
      private static final String HELP_TWO_DASH = "--help";
      private static final String CMD = "VM.metaspace";
+     private static final String ILLEGAL = "IllegalArgumentException: Unknown argument";

      public static void main(String[] args) throws Exception {

          // Sanity check with empty input
          OutputAnalyzer output = JcmdBase.jcmd();
***************
*** 59,68 ****
--- 60,72 ----
          testIgnoreAdditionalArgs(HELP_ONE_DASH, expectedOutput);
          testIgnoreAdditionalArgs(HELP_TWO_DASH, expectedOutput);

          testIgnoreTrailingSpaces(HELP_ONE_DASH, expectedOutput);
          testIgnoreTrailingSpaces(HELP_TWO_DASH, expectedOutput);
+
+         testSimilarCommand(HELP_ONE_DASH + "ello", ILLEGAL);
+         testSimilarCommand(HELP_TWO_DASH + "me", ILLEGAL);
      }

      private static void testExpectedUsage(String helpOption, String expectedOutput) throws Exception {
          verifyOutput(new String[] {CMD, helpOption}, expectedOutput,
                  "Expected jcmd to accept '%s' suboption as a command argument and issue the same help output.".formatted(helpOption));
***************
*** 76,93 ****
--- 80,114 ----
      private static void testIgnoreTrailingSpaces(String helpOption, String expectedOutput) throws Exception {
          verifyOutput(new String[] {CMD, "%s    ".formatted(helpOption)}, expectedOutput,
                  "Expected jcmd to accept '%s' suboption with trailing spaces".formatted(helpOption));
      }

+     private static void testSimilarCommand(String helpOption, String expectedOutput) throws Exception {
+         verifyOutputContains(new String[] {CMD, helpOption}, expectedOutput,
+                 "Expected jcmd to NOT accept '%s' suboption with trailing content".formatted(helpOption));
+     }
+
      private static void verifyOutput(String[] args, String expectedOutput, String errorMessage) throws Exception {
          OutputAnalyzer output = JcmdBase.jcmd(args);
          String issuedOutput = output.getOutput();
          if (!expectedOutput.equals(issuedOutput)) {
              System.out.println("Expected output: ");
              System.out.println(expectedOutput);
              System.out.println("Issued output: ");
              System.out.println(issuedOutput);
              throw new Exception(errorMessage);
+         }
+     }
+
+     private static void verifyOutputContains(String[] args, String expectedOutput, String errorMessage) throws Exception {
+         OutputAnalyzer output = JcmdBase.jcmd(args);
+         String issuedOutput = output.getOutput();
+         if (!issuedOutput.contains(expectedOutput)) {
+             System.out.println("Expected output: ");
+             System.out.println(expectedOutput);
+             System.out.println("Issued output: ");
+             System.out.println(issuedOutput);
+             throw new Exception(errorMessage);
          }
      }
  }
```

**Comment by SoniaZaldana** [comment]
Hi @kevinjwalls, thanks for taking a look! I made some updates based on your comments.

**Comment by kevinjwalls** [comment]
One more thing from me, as as_string() uses the ResourceArea, it probably makes sense to move the ResourceMark at line 412 up to line 403.

OR, avoid the allocation, we are definitely just reading the characters, so we could use: 
  char *rest = (char*) args.base();
(strtok_r wants a non-const char*)
  
Then OK great, done.

If you are making that update and wanted to get the .cpp file comment at line 403 on just two lines, and in the test if you wanted to call the "-h" something other than "POSIX", those would not be bad final updates from my point of view.

File: src/hotspot/share/services/diagnosticFramework.cpp [hunk 1]
```diff
@@ -411,6 +420,25 @@ void DCmd::parse_and_execute(DCmdSource source, outputStream* out,
   }
 }
 
+bool DCmd::reorder_help_cmd(CmdLine line, stringStream &updated_line) {
+  stringStream args;
+  args.print("%s", line.args_addr());
+  char *rest = args.as_string();
+  char *token = strtok_r(rest, " ", &rest);
```
Comment (tstuefe) [inline src/hotspot/share/services/diagnosticFramework.cpp:427]: Nit, please use char* V not char *V, the former is much more common in hotspot.

**Comment by tstuefe** [comment]
> OR, avoid the allocation, we are definitely just reading the characters, so we could use: char _rest = (char_) args.base(); (strtok_r wants a non-const char*)
> 
> Then OK great, done.
> 

No, please don't do that. It is bad practice - don't get in the habit of casting away constness unless you are sure you know what happens. strtok modifies the string, that is why it needs write access, and we need to preserve CmdLine for the case when its *not* a help command.

Other than that, it looks okay for me and I will approve it in this form. However, if you have the time, you could rewrite the reorder function to not use strtok but to do the scanning yourself.

- with a pointer pointing to the start of arguments
- walk pointer for as long as there are spaces, or until \0
- then, use strncmp (notice the n) to check for the commands
- then you are done. No need to copy the string, no need for strtok to filet the string in its whole glory just for most of it to be ignored later. 

Cheers, Thomas

**Review by tstuefe — APPROVED** [review]

**Review by kevinjwalls — APPROVED** [review]
Looks good to me, I think it's done, thanks for your patience.

**Comment by SoniaZaldana** [comment]
Thank you both for the reviews! @tstuefe @kevinjwalls

**Pull request closed without merging**